### ISSUE TITLE
Return Nil For Invalid Slice Index

### ISSUE THREAD

**Issue by cwarden** [antonmedv/expr#341]
I'm looking for a nicer way of having an expression return nil when referencing an invalid array index.  For example, `filter(products, .Name matches "NoMatch")[0]` results in an "slice index out of range" error.

I'm trying to return first-match-or-nil and looking a simpler expression than
`none(products, .Name matches "A") ? nil : filter(products, .Name matches "A")[0]`

**Comment by antonmedv** [comment]
I have a few ideas. First, I’m working on adding variables. So something like this will be available:

    let arr = filter(products, .Name matches "NoMatch")
    len(arr) > 0 ? arr[0] : nil

Another thing is chains operator. Expr already has .? operator. So maybe adding ?.[] will make sense. 

    filter(products, .Name matches "NoMatch")?.[0]
    
I like your clever solution using none)

**Comment by antonmedv** [comment]
Or maybe adding a special function first() makes sense. 

    first(filter(products, .Name matches "NoMatch"))

**Comment by cwarden** [comment]
Maybe an invalid index should return nil.  `filter(products, .Name matches "NoMatch")[:1]` returns an empty slice rather than an out of range error.

**Comment by antonmedv** [comment]
Expr works as Python. 

Slice operation always return a slice. Index operation returns anything. 

Consider and array this nil:

    [nil, nil, 1]
    
We can’t distinguish nil in array vs no index, if we going to return nil. 

Let’s research more on the issue. I agree what some better solutions for getting values from arrays should exist.

**Comment by antonmedv** [comment]
For example in your case:

    none(products, .Name matches "A") ? nil : filter(products, .Name matches "A")[0]
    
Lets add find function 

    find(products, .Name matches "A")

will return nil if not found.

**Comment by antonmedv** [comment]
Yes, probably not important. But now changing behavior will definitely break someone something.

**Comment by antonmedv** [comment]
Kotlin has [getOrNull](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/get-or-null.html):

```kotlin
val list = listOf(1, 2, 3)
println(list.getOrNull(0)) // 1
println(list.getOrNull(2)) // 3
println(list.getOrNull(3)) // null
```

**Comment by antonmedv** [comment]
Maybe something like get, first, last?
```
filter(products, .Name matches "NoMatch") | get(0)
```

**Comment by antonmedv** [comment]
Done!
```
	{`first(ArrayOfString)`, "foo"},
	{`first(ArrayOfInt)`, 1},
	{`first(ArrayOfAny)`, 1},
	{`last(ArrayOfString)`, "baz"},
	{`last(ArrayOfInt)`, 3},
	{`last(ArrayOfAny)`, true},
	{`get(ArrayOfString, 1)`, "bar"},
	{`get(ArrayOfString, 99)`, nil},
	{`get(ArrayOfInt, 1)`, 2},
	{`get(ArrayOfInt, -1)`, 3},
	{`get(ArrayOfAny, 1)`, "2"},
	{`get({foo: 1, bar: 2}, "foo")`, 1},
	{`get({foo: 1, bar: 2}, "unknown")`, nil},
```
